### DISCUSSION TITLE
Support for Opening External Terminal Window like Warp or iTerm2

### DISCUSSION THREAD

**Discussion by rommyarb** [zed-industries/zed#7974] [Feature Requests]
I've been exploring Zed and appreciate its seamless integration with the user's default shell. I wondered if we could add support for opening external terminal windows (not in the dock) like [Warp](https://www.warp.dev) or [iTerm2](https://iterm2.com), by customizing the default key binding or something else.

VSCode has this feature on its `settings.json`:
```json
{
 "terminal.external.osxExec": "Warp.app",
}
```

This could offer more flexibility for users who prefer or require different terminal environments for their projects. What do you guys think about this?

**Comment by peterpeterparker** [comment]
That would be nice. With JetBrains editors on Mac, I like being able to drag and drop a project over the terminal icon to open the related folder.

![dragdrop](https://github.com/zed-industries/zed/assets/16886711/91ca8b23-0a6d-418d-8668-afceaca5f5ca)

**Comment by marlonmp** [comment]
There is a way to open the external terminal with a custom key binding, it is not the same as VSCode, but it works for me.

First, you have to add a task in your Zed tasks file, in my case I'm using gnome-terminal

```json
[
  ...
  {
    "label": "external-terminal",
    "command": "gnome-terminal",
    "args": ["--working-directory=$ZED_WORKTREE_ROOT"],
    "reveal": "never",
    "hide": "always",
    "shell": "system"
  }
]
```

But If you are using Warp in linux

```json
  {
    "label": "external-terminal",
    "command": "nohup",
    "args": ["warp-terminal warp://action/new_tab?path=$ZED_WORKTREE_ROOT"],
    "reveal": "never",
    "hide": "always",
    "shell": "system"
  }

```

Or if you are using Warp in Mac, I think this will work

```json
  {
    "label": "external-terminal",
    "command": "open",
    "args": ["\"[warp://action/new_tab?path=$ZED_WORKTREE_ROOT\""],
    "reveal": "never",
    "hide": "always",
    "shell": "system"
  }

```

Once you add the task you can spawn the task to check if it works.

Then you can create your custom key binding to spawn your task like this:

```json
[
  ...
  {
    "context": "Workspace",
    "bindings": {
      "ctrl-alt-c": ["task::Spawn", { "task_name": "external-terminal" }]
    }
  },
]
```

And thats all, I hope this solution works for you while Zed adds support for this feature.

**Reply by dijalmass** [reply]
Worked for me, thanks:

```json
{
  "label": "external-terminal",
  "command": "flatpak run com.raggesilver.BlackBox",
  "args": ["--working-directory", "$ZED_WORKTREE_ROOT"],
  "reveal": "never",
  "hide": "always",
  "shell": "system"
}
```

And I created as well the shortcut, but the only problem is if i press again the shortcut with one spawn already opened, it will be closed and opened other. 

But for this just open other tabs in the terminal  :smiley:

**Reply by treyg** [reply]
For some reason I had to give it the full path to `/usr/bin/open` to get it to work for ghostty on mac, but this works. It'll open the parent folder of whatever file you're viewing: 
```json
{
  "label": "external-ghostty",
  "command": "/usr/bin/open -a Ghostty \"$ZED_DIRNAME\"",
  "reveal": "never",
  "hide": "always",
}
```

**Comment by CompaCodeRC** [comment]
I did it this way for Warp:

Create a folder for my script:
```mkdir ~/scripts```

Enter:
```cd ~/scripts```

Create a file:
```nano launch-warp.sh```

Paste, save and exit:
```
#!/bin/bash
open -a "Warp" "$PWD"
```

I gave permissions:
```chmod +x ~/scripts/launch-warp.sh```

Add in settings.json (zed):
```
"terminal": {
    "shell": {
      "with_arguments": {
        "program": "/bin/bash",
        "args": ["-c", "~/scripts/launch-warp.sh"]
      }
    }
  }
```
  
Ready, now the terminal opens when you open the project or click on the terminal icon or with the commands to open terminal

**Reply by joaquingatica** [reply]
@CompaCodeRC thank you for this! I ended up using a variant of it that skips the need for an external script. It's also adapted for `iTerm`, but same idea. 

```json
  "terminal": {
    "shell": {
      "with_arguments": {
        "program": "/bin/bash",
        "args": ["-c", "open -a \"iTerm\" \"$PWD\""]
      }
    }
  }
```

**Reply by rolodato** [reply]
You can also avoid spawning a bash process by running `open` directly:

```json
  "terminal": {
    "shell": {
      "with_arguments": {
        "program": "/usr/bin/open",
        "args": ["-a", "iTerm", "."]
      }
    }
  }
```

**Comment by JosephTLyons** [comment]
Thanks for reaching out. We want to better support our community, and the best way for us to do that is through our official Discord. The team is active there and happy to help.

We’re closing the “Help and General Q&A” category on GitHub Discussions and directing all support questions to Discord going forward.

https://zed.dev/community-links
